### PR TITLE
Issue #50 - Support octal/hex escape sequences in strings

### PR DESCRIPTION
**Pull request by christian-stephen**

### CODE REVIEW

**Comment by ShivamSarodia** [comment]
Hi Christian, thank you for this PR! Couple things:

- Feel free to add yourself as a contributor on the README
- Could you please write some tests for this added functionality? You can add them to tests/feature_tests/string.c.

Thanks again! :)

File: shivyc/lexer.py [hunk 1]
```diff
@@ -321,11 +323,32 @@ def read_string(line, start, delim, null):
...
+              and line[i + 1].c in escapes):
             chars.append(escapes[line[i + 1].c])
             i += 2
+        elif (i + 1 < len(line)
+              and line[i].c == "\\"
+              and line[i + 1].c in octdigits):
+            octal = line[i + 1].c
+            i += 2
+            while (i < len(line)
+                   and len(octal) < 3
```
Comment (ShivamSarodia) [inline shivyc/lexer.py]: Good catch with the `len(octal) < 3`

File: shivyc/lexer.py [hunk 2]
```diff
@@ -321,11 +323,32 @@ def read_string(line, start, delim, null):
...
+                octal += line[i].c
+                i += 1
+            chars.append(int(octal, 8))
+        elif (i + 2 < len(line)
+              and line[i].c == "\\"
+              and line[i + 1].c == "x"
+              and line[i + 2].c in hexdigits):
+            hexa = line[i + 2].c
+            i += 3
+            if i < len(line) and line[i].c in hexdigits:
```
Comment (ShivamSarodia) [inline shivyc/lexer.py]: I believe you meant this to be a `while` statement like above. In the C11 spec 6.4.4.4.1 (page 68), a `hexadecimal-escape-sequence` is `\x hexadecimal-digit` or `hexadecimal-escape-sequence hexadecimal-digit`. This means that in a `hexadecimal-escape-sequence`, `\x` can be followed by arbitrarily many `hexadecimal-digit` by applying the second rule repeatedly.

**Comment by christian-stephen** [comment]
Thanks Shivam. I made the adjustments you suggested.

**Comment by ShivamSarodia** [comment]
Thanks again! If you'd like to continue your contributions to ShivyC, I'll be keeping the Issues tab updated with pending tasks. Feel free to pick up any that look interesting - if you have questions or thoughts, you can comment on the issue :)

**Pull request merged**